test/UnitTestHelpers.hs

```diff
@@ -0,0 +1,362 @@
+-- component of the library, this implementation can be substituted.
+module UnitTestHelpers
+  ( emptyCliState
+  , executorCli
+  , stubLevel
+  , stubState
+  , stubCliState
+  , testActor
+  , testActorId
+  , testActorWithItem
+  , testCliStateWithItem
+  , testFactionId
+  , testItemId
+  , testItemKind
+  , testLevelId
+-- #ifdef EXPOSE_INTERNAL
+--     -- * Internal operations
+  , CliMock(..)
+  , CliState(..)
+-- #endif
+  ) where
+
+import Prelude ()
+
+import Game.LambdaHack.Core.Prelude
+
+import qualified Control.Monad.IO.Class as IO
+import System.IO (putStr)
+
+import Control.Monad.Trans.State.Strict
+    ( StateT(StateT, runStateT), gets, state, evalStateT )
+
+import qualified Data.EnumMap.Strict as EM
+
+import           Game.LambdaHack.Atomic (MonadStateWrite (..))
+import           Game.LambdaHack.Client
+import qualified Game.LambdaHack.Client.BfsM as BfsM
+import           Game.LambdaHack.Client.HandleResponseM
+import           Game.LambdaHack.Client.MonadClient
+import           Game.LambdaHack.Client.State
+import           Game.LambdaHack.Client.UI
+import           Game.LambdaHack.Client.UI.ActorUI
+import           Game.LambdaHack.Client.UI.ContentClientUI
+import           Game.LambdaHack.Client.UI.Content.Screen
+import           Game.LambdaHack.Client.UI.Frontend
+import           Game.LambdaHack.Client.UI.Key (KMP (..))
+import qualified Game.LambdaHack.Client.UI.Key as K
+import           Game.LambdaHack.Client.UI.PointUI
+import           Game.LambdaHack.Client.UI.SessionUI
+import           Game.LambdaHack.Client.UI.UIOptions
+
+import           Game.LambdaHack.Common.Actor
+import           Game.LambdaHack.Common.Area
+import           Game.LambdaHack.Common.ClientOptions
+import           Game.LambdaHack.Common.Faction
+import           Game.LambdaHack.Common.Kind
+import           Game.LambdaHack.Common.Level
+import           Game.LambdaHack.Common.MonadStateRead
+import           Game.LambdaHack.Common.Perception
+import           Game.LambdaHack.Common.Point
+import           Game.LambdaHack.Common.State
+import           Game.LambdaHack.Common.Time
+import           Game.LambdaHack.Common.Types
+
+import           Game.LambdaHack.Content.ItemKind
+import           Game.LambdaHack.Content.RuleKind
+import           Game.LambdaHack.Content.TileKind
+import qualified Game.LambdaHack.Core.Dice as Dice
+
+import qualified Game.LambdaHack.Definition.Ability as Ability
+import           Game.LambdaHack.Definition.Color
+import           Game.LambdaHack.Definition.Flavour
+
+import Content.FactionKind
+
+import Game.LambdaHack.Common.Misc
+
+
+-- Read UI requests from the client and send them to the frontend,
+fchanFrontendStub :: ChanFrontend
+fchanFrontendStub =
+  ChanFrontend $ \case
+    FrontFrame _ -> putStr "FrontFrame"
+    FrontDelay _ -> putStr "FrontDelay"
+    FrontKey _ _ -> do return KMP { kmpKeyMod=K.escKM, kmpPointer=PointUI 0 0 }
```

Comment:
this `do` is spurious; you only need it if you sequence many monadic expressions, not a single one